### ISSUE TITLE
Linux arm32 GCStress=C test failures

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/runtime#31770]
https://dev.azure.com/dnceng/public/_build/results?buildId=507060&view=results

Many tests fail with gcstress0xc_jitstress1, gcstress0x_jitstress2, gcstress0xc_zapdisable_jitstress2.

E.g., `JIT/HardwareIntrinsics/General/Vector256_1/Vector256_1_ro/Vector256_1_ro.sh`:

`
\nAssert failure(PID 55 [0x00000037], Thread: 55 [0x0037]): SanityCheck()\n File: /__w/8/s/src/coreclr/src/vm/methodtable.cpp Line: 9842\n Image: /root/helix/work/correlation/corerun\n\nptrace(GETFPREGS, 55) FAILED 5 (Input/output error)\n/root/helix/work/workitem/JIT/HardwareIntrinsics/General/Vector256_1/Vector256_1_ro/Vector256_1_ro.sh: line 275: 55 Aborted $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"\n\nReturn code: 1\nRaw output file: /root/helix/work/workitem/JIT/HardwareIntrinsics/Reports/JIT.HardwareIntrinsics/General/Vector256_1/Vector256_1_ro/Vector256_1_ro.output.txt\nRaw output:\nBEGIN EXECUTION\n/root/helix/work/correlation/corerun Vector256_1_ro.dll ''\nBeginning test case Zero.Byte at 2/4/2020 8:54:38 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:43 PM\nBeginning test case Zero.Double at 2/4/2020 8:54:43 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:44 PM\nBeginning test case Zero.Int16 at 2/4/2020 8:54:44 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:44 PM\nBeginning test case Zero.Int32 at 2/4/2020 8:54:44 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:44 PM\nBeginning test case Zero.Int64 at 2/4/2020 8:54:44 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:44 PM\nBeginning test case Zero.SByte at 2/4/2020 8:54:44 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:44 PM\nBeginning test case Zero.Single at 2/4/2020 8:54:44 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:44 PM\nBeginning test case Zero.UInt16 at 2/4/2020 8:54:44 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:44 PM\nBeginning test case Zero.UInt32 at 2/4/2020 8:54:44 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:44 PM\nBeginning test case Zero.UInt64 at 2/4/2020 8:54:44 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:45 PM\nBeginning test case As.Byte at 2/4/2020 8:54:45 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunGenericScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:49 PM\nBeginning test case As.Double at 2/4/2020 8:54:49 PM\nRandom seed: 20010415; set environment variable CORECLR_SEED to this value to repro\n\nBeginning scenario: RunBasicScenario\nBeginning scenario: RunGenericScenario\nBeginning scenario: RunReflectionScenario\n\nEnding test case at 2/4/2020 8:54:51 PM\nBeginni
`

Also:
```
JIT/HardwareIntrinsics/General/Vector64_1/Vector64_1_ro/Vector64_1_ro.sh
```

Another failure mode:
`
\nAssert failure(PID 642 [0x00000282], Thread: 642 [0x0282]): !CREATE_CHECK_STRING(pMT && pMT->Validate())\n File: /__w/8/s/src/coreclr/src/vm/object.cpp Line: 557\n Image: /root/helix/work/correlation/corerun\n\nptrace(GETFPREGS, 642) FAILED 5 (Input/output error)\n/root/helix/work/workitem/Interop/PInvoke/SizeParamIndex/PInvoke/PassingByRef/PassingByRefTest/PassingByRefTest.sh: line 275: 642 Aborted $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"\n\nReturn code: 1\nRaw output file: /root/helix/work/workitem/Interop/PInvoke/Reports/Interop.PInvoke/SizeParamIndex/PInvoke/PassingByRef/PassingByRefTest/PassingByRefTest.output.txt\nRaw output:\nBEGIN EXECUTION\n/root/helix/work/correlation/corerun PassingByRefTest.dll ''\n\nScenario(byte==>BYTE):Array_Size(M->N)=1,Array_Size(N->M)= byte.MinValue Starts!\nScenario(byte==>BYTE):Array_Size(M->N)=1,Array_Size(N->M)= byte.MinValue Ends!\n\nScenario(sbyte==>CHAR): Array_Size(M->N) = 10, Array_Size(N->M) = sbyte.Max Starts!\nScenario(sbyte==>CHAR): Array_Size(M->N) = 10, Array_Size(N->M) = sbyte.Max Ends!\n\nScenario(short==>SHORT)1: Array_Size(M->N) = -1, Array_Size(N->M) = 20 Starts!\nScenario(short==>SHORT)1: Array_Size(M->N) = -1, Array_Size(N->M) = 20 Ends!\n\nScenario(short==>SHORT)2: Array_Size(M->N) = 10, Array_Size(N->M) = -1 Starts!\nScenario(short==>SHORT)2: Array_Size(M->N) = 10, Array_Size(N->M) = -1 Ends!\n\nScenario(ushort==>USHORT): Array_Size(M->N) = 0, Array_Size(N->M) = ushort.MaxValue Starts!\nScenario(ushort==>USHORT): Array_Size(M->N) = 0, Array_Size(N->M) = ushort.MaxValue Ends!\n\nScenario(Int32==>LONG):Array_Size(M->N)=10, Array_Size(N->M)=1 Starts!\nScenario(Int32==>LONG):Array_Size(M->N)=10, Array_Size(N->M)=1 Ends!\n\nScenario(UInt32==>ULONG):Array_Size(M->N)=1234,Array_Size(N->M)=4321 Starts!\nScenario(UInt32==>ULONG):Array_Size(M->N)=1234,Array_Size(N->M)=4321 Ends!\n\nScenario(long==>LONGLONG):Array_Size(M->N)=10,Array_Size(N->M)=20 Starts!\nScenario(long==>LONGLONG):Array_Size(M->N)=10,Array_Size(N->M)=20 Ends!\n\nScenario(ulong==>ULONGLONG):Array_Size(M->N)=0, Array_Size(N->M)=0 Starts!\nScenario(ulong==>ULONGLONG):Array_Size(M->N)=0, Array_Size(N->M)=0 Ends!\nWriting minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.642.dmp\nExpected: 100\nActual: 134\nEND EXECUTION - FAILED\nTest Harness Exitcode is : 1\nTo run the test:\n> set CORE_ROOT=/root/helix/work/correlation\n> /root/helix/work/workitem/Interop/PInvoke/SizeParamIndex/PInvoke/PassingByRef/PassingByRefTest/PassingByRefTest.sh\nExpected: True\nActual: False
`

with:
```
Interop/PInvoke/SizeParamIndex/PInvoke/PassingByRef/PassingByRefTest/PassingByRefTest.sh
Interop/PInvoke/SizeParamIndex/PInvoke/PassingByOut/PassingByOutTest/PassingByOutTest.sh
Interop/PInvoke/Array/MarshalArrayAsParam/AsDefault/AsDefaultTest/AsDefaultTest.sh
Interop/PInvoke/Array/MarshalArrayAsParam/AsLPArray/AsLPArrayTest/AsLPArrayTest.sh
Interop/PInvoke/SizeParamIndex/ReversePInvoke/PassingByRef/PassingByRefTest/PassingByRefTest.sh
Interop/StructPacking/StructPacking/StructPacking.sh
Interop/PInvoke/Decimal/DecimalTest/DecimalTest.sh
Interop/PrimitiveMarshalling/Bool/BoolTest/BoolTest.sh

CoreMangLib/system/runtime/interopservices/marshal/MarshalSizeOf2_PSC/MarshalSizeOf2_PSC.sh
```

Another failure mode:
`
ptrace(GETFPREGS, 2530) FAILED 5 (Input/output error)\n/root/helix/work/workitem/JIT/Methodical/tailcall/_il_dbgdeep_array/_il_dbgdeep_array.sh: line 275: 2530 Segmentation fault $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"\n\nReturn code: 1\nRaw output file: /root/helix/work/workitem/JIT/Methodical/Reports/JIT.Methodical/tailcall/_il_dbgdeep_array/_il_dbgdeep_array.output.txt\nRaw output:\nBEGIN EXECUTION\n/root/helix/work/correlation/corerun _il_dbgdeep_array.dll ''\nWriting minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.2530.dmp\nExpected: 100\nActual: 139\nEND EXECUTION - FAILED\nTest Harness Exitcode is : 1\nTo run the test:\n> set CORE_ROOT=/root/helix/work/correlation\n> /root/helix/work/workitem/JIT/Methodical/tailcall/_il_dbgdeep_array/_il_dbgdeep_array.sh\nExpected: True\nActual: False
`
with:
```
JIT/Methodical/tailcall/_il_dbgdeep_array/_il_dbgdeep_array.sh
```
and similar tests.

category:correctness
theme:testing
skill-level:expert
cost:medium

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
These look similar to the set that was failing on windows recently (#1971), so perhaps dodgy gc liveness around late inserted code is the culprit; see also #2104.

**Comment by AndyAyersMS** [comment]
I'll investigate.

**Comment by AndyAyersMS** [comment]
I can repro, but it's intermittent, maybe one in 5 runs? So likely a missing GC root from jitted code.

Looking at `Vector64_1_ro`, the issue seems to be that there's a special contract with the leave profiler hook to transfer the return value from R0 to R2 and back around the call, and we don't seem to get proper GC updates.

So when a method like `Vector64:ToVector128` is invoked via reflection and the hidden return value buffer  pointer is in R2, it can be trashed by GC.

The code looks like it is trying to do the right thing; not yet clear why it doesn't work.
```
Generating: N046 (  3,  2) [000016] -------N-----       t16 =    LCL_VAR   byref  V04 tmp1          r0 REG r0
IN001c:         2B      ldr     r0, [sp+0x04]	// [V04 tmp1]
							Byref regs: 0000 {} => 0001 {r0}
                                                               /--*  t16    byref  
Generating: N048 (  4,  3) [000017] -------------             *  RETURN    byref  REG NA
							Byref regs: 0001 {r0} => 0000 {}
							Byref regs: 0000 {} => 0001 {r0}
      G_M64667_IG29:        ; offs=000044H, funclet=00, bbWeight=1   
IN001d:         2B      mov     r2, r0
							Byref regs: 0001 {r0} => 0005 {r0 r2}

      G_M64667_IG30:        ; offs=000046H, funclet=00, bbWeight=1   
IN001e:         4B      movw    r0, 0x5aff

      G_M64667_IG31:        ; offs=000048H, funclet=00, bbWeight=1   
IN001f:         4B      movt    r0, 0xe1fd
							Byref regs: 0005 {r0 r2} => 0004 {r2}

      G_M64667_IG32:        ; offs=00004CH, funclet=00, bbWeight=1   
IN0020:         4B      movw    r12, 0x5aff

      G_M64667_IG33:        ; offs=000050H, funclet=00, bbWeight=1   
IN0021:         4B      movt    r12, 0xe1fd
NoGC Call: savedSet=57FF {r0 r1 r2 r3 r4 r5 r6 r7 r8 r9 r10 r12 lr}
Call: GCvars= {}, gcrefRegs=0000 {}, byrefRegs=0004 {r2}

      G_M64667_IG34:        ; offs=000054H, funclet=00, bbWeight=1   
[34] Rec call GC vars = 
IN0022:         2B      blx     r12		// CORINFO_HELP_PROF_FCN_LEAVE

      G_M64667_IG35:        ; offs=000058H, funclet=00, bbWeight=1   
IN0023:         2B      mov     r0, r2
							Byref regs: 0004 {r2} => 0005 {r0 r2}
							Byref regs: 0005 {r0 r2} => 0001 {r0}
							Byref regs: 0001 {r0} => 0000 {}

;;; and then later r2 comes live too late...

New byrReg live regs=0000 {}
byrReg -[r0]
byrReg +[r0]
IN001c: 000064  9801           ldr     r0, [sp+0x04]	// [V04 tmp1]
						;; bbWeight=1    PerfScore 1.00
G_M64667_IG30:        ; func=00, offs=000066H, size=0002H, extend
IN001d: 000066  4602           mov     r2, r0
						;; bbWeight=1    PerfScore 1.00
G_M64667_IG31:        ; func=00, offs=000068H, size=0004H, extend
byrReg -[r0]
IN001e: 000068  F645 20FF      movw    r0, 0x5aff
						;; bbWeight=1    PerfScore 1.00
G_M64667_IG32:        ; func=00, offs=00006CH, size=0004H, extend
IN001f: 00006C  F2CE 10FD      movt    r0, 0xe1fd
						;; bbWeight=1    PerfScore 1.00
G_M64667_IG33:        ; func=00, offs=000070H, size=0004H, extend
IN0020: 000070  F645 2CFF      movw    r12, 0x5aff
						;; bbWeight=1    PerfScore 1.00
G_M64667_IG34:        ; func=00, offs=000074H, size=0004H, extend
IN0021: 000074  F2CE 1CFD      movt    r12, 0xe1fd
						;; bbWeight=1    PerfScore 1.00
G_M64667_IG35:        ; func=00, offs=000078H, size=0002H, extend
New byrReg live regs=0004 {r2}
byrReg +[r2]
IN0022: 000078  47E0           blx     r12		// CORINFO_HELP_PROF_FCN_LEAVE
						;; bbWeight=1    PerfScore 1.00
G_M64667_IG36:        ; func=00, offs=00007AH, size=0002H, extend
IN0023: 00007A  4610           mov     r0, r2
						;; bbWeight=1    PerfScore 1.00
G_M64667_IG37:        ; func=00, offs=00007CH, size=000AH, epilog, nogc, extend
IN002f: 00007C  B006           add     sp, 24
IN0030: 00007E  E8BD 4C10      pop     {r4,r10,r11,lr}
IN0031: 000082  B004           add     sp, 16
IN0032: 000084  4770           bx      lr
```

**Comment by AndyAyersMS** [comment]
Ok, figured that part out -- we weren't passing EA_BYREF to `emitIns_R_R` for the hidden return buffer case.

There is also potentially an issue with arm's `genProfilingLeaveCallback` checking `compRetType` instead of `compRetNativeType`, so I'll fix that too.

Let me see if any of these other failures repro after the above.

**Comment by AndyAyersMS** [comment]
For `_il_dbgdeep_array`, the issue is that `R0` is not the return value, if we're in a jump or tail call return. So deducing gc-liveness of `R0` based on return value characteristics of the method can't work.

For the jump case `R0` is dead, as it gets spilled via `genJmpMethod`. Still looking at what happens in the tail call case...

**Comment by AndyAyersMS** [comment]
Should be fixed by #32674.
